Api authentication

bagatka opened the pull request.

andrei-kruhliakou-itechart commented on iTechArt.CinemaWebApp.API/Application/Services/AccountService.cs:
```diff
@@ -0,0 +1,112 @@
+                return new Response<LoginResponse>($"No Accounts Registered with {request.Email}.");
+            }
+
+            var validatePassword = BCrypter.EnhancedVerify(request.Password, user.PasswordHash);
```
Shouldn't you check that password confirmation is equal to password?

bagatka replied:
`validatePassword` is a bool variable, so after this line I made a check
`if (!validatePassword) return new Response<LoginResponse>($"Invalid Credentials for {request.Email}.");`

Maybe there is a problem with naming of `validatePassword` variable

andrei-kruhliakou-itechart replied:
Sorry, this comment was for Register method. You have password and password confirmation fields in the request but use only one of them.

andrei-kruhliakou-itechart commented on iTechArt.CinemaWebApp.API/Controllers/AccountController.cs:
```diff
@@ -0,0 +1,33 @@
+        [HttpPost("login")]
+        public async Task<IActionResult> LoginAsync(LoginRequest request)
+        {
+            return Ok(await _accountService.LoginAsync(request));
```
You might want to check model state before passing data to service.

andrei-kruhliakou-itechart commented on iTechArt.CinemaWebApp.API/Application/Wrappers/Response.cs:
```diff
@@ -0,0 +1,23 @@
+{
+    public class Response<T>
+    {
+        public bool Succeeded { get; set; }
+        public string Message { get; set; }
+        public T Data { get; set; }
+        public Response()
+        {
+        }
+        public Response(T data, string message = null)
+        {
+            Succeeded = true;
+            Message = message;
+            Data = data;
+        }
+        public Response(string message)
+        {
+            Succeeded = false;
+            Message = message;
+        }
```
Please separate each method and property with empty line

andrei-kruhliakou-itechart commented on iTechArt.CinemaWebApp.API/Application/Services/AccountService.cs:
```diff
@@ -0,0 +1,112 @@
+            _mapper = mapper;
+            _config = configuration;
+        }
+        public async Task<Response<LoginResponse>> LoginAsync(LoginRequest request)
```
Separate methods please

andrei-kruhliakou-itechart commented on iTechArt.CinemaWebApp.API/Application/Services/AccountService.cs:
```diff
@@ -0,0 +1,112 @@
+        {
+            var user = await _context.Users.FirstOrDefaultAsync(user => user.Email == request.Email);
+
+            if(user == null)
```
Space after if

andrei-kruhliakou-itechart commented on iTechArt.CinemaWebApp.API/Application/Services/AccountService.cs:
```diff
@@ -0,0 +1,112 @@
+
+            var validatePassword = BCrypter.EnhancedVerify(request.Password, user.PasswordHash);
+            
+            if(!validatePassword)
```
Space after if

andrei-kruhliakou-itechart commented on iTechArt.CinemaWebApp.API/Application/Services/AccountService.cs:
```diff
@@ -0,0 +1,112 @@
+
+            return new Response<LoginResponse>(response, $"{request.Email} has been successfully registered.");
+        }
+        private JwtSecurityToken GenerateJWTToken(User userInfo)
```
Separate methods please

bagatka commented:
I've fixed all if and separating line problems.
Also removed unused imports.

Btw is there any rule how to place using's? I mean the order of placing them. Rider places System imports first, then my own files imports, then Microsoft and all others. I think Microsoft imports should be "higher" than my own

andrei-kruhliakou-itechart commented on iTechArt.CinemaWebApp.API/Application/Services/AccountService.cs:
```diff
@@ -33,44 +33,38 @@ public class AccountService
-            {
-                return new Response<LoginResponse>($"No Accounts Registered with {request.Email}.");
-            }
+            if (user == null) return new Response<LoginResponse>($"No Accounts Registered with {request.Email}.");
```
Please always code blocks with if, loops, etc.

andrei-kruhliakou-itechart commented on iTechArt.CinemaWebApp.API/Application/Services/AccountService.cs:
```diff
@@ -33,44 +33,38 @@ public class AccountService
-            {
-                return new Response<LoginResponse>($"Invalid Credentials for {request.Email}.");
-            }
+            var validatePassword = BCrypter.EnhancedVerify(request.Password,
+                user.PasswordHash);
```
Either use one line (if it doesn't exceed 120 characters) or the following style:
```suggestion
            var validatePassword = BCrypter.EnhancedVerify(
                request.Password,
                user.PasswordHash
            );
```

andrei-kruhliakou-itechart commented on iTechArt.CinemaWebApp.API/Controllers/FilmsController.cs:
```diff
@@ -47,42 +42,33 @@ public async Task<ActionResult<Film>> GetFilmById(int id)
             await _context.SaveChangesAsync();
 
-            return CreatedAtAction(nameof(GetFilmById), new { id = newFilm.Id }, newFilm);
+            return CreatedAtAction(nameof(GetFilmById), new {id = newFilm.Id}, newFilm);
```
Previous formatting (with spaces) is preferred.

andrei-kruhliakou-itechart commented on iTechArt.CinemaWebApp.API/Controllers/AccountController.cs:
```diff
@@ -21,13 +21,15 @@ public AccountController(AccountService accountService)
         [HttpPost("login")]
         public async Task<IActionResult> LoginAsync(LoginRequest request)
         {
+            if (!ModelState.IsValid) return BadRequest("Incorrect request body.");
```
You can be more specific with error message. Use information from model state.

bagatka replied:
# What do you think about this solution?

### Creating a custom attribute:

![image](https://user-images.githubusercontent.com/45293863/94903093-ec8f6b80-04a1-11eb-80ff-d44e38d9513b.png)


### In controller:

![image](https://user-images.githubusercontent.com/45293863/94903235-295b6280-04a2-11eb-8c69-b0f8f332cc8d.png)


### In Client App:
#### Every time when the model is invalid controller will automatically send BadRequestObjectResult:

![image](https://user-images.githubusercontent.com/45293863/94903329-4d1ea880-04a2-11eb-95e5-84faba84a74b.png)

bagatka commented:
So, validation is working fine. Client App or Postman receives error if something from viewmodel is incorrect. Error message is automatically applies to response.
I don't have to compare `Password` and `Confirm password` 'cause I've used `[Compare("Password")]` in model.

andrei-kruhliakou-itechart commented on iTechArt.CinemaWebApp.API/Application/Services/AccountService.cs:
```diff
@@ -0,0 +1,121 @@
+
+            var claims = new[]
+            {
+                new Claim(JwtRegisteredClaimNames.Sub,
+                    userInfo.UserName),
+                new Claim("firstName",
+                    userInfo.FirstName),
+                new Claim("lastName",
+                    userInfo.LastName),
+                new Claim("role",
+                    userInfo.Role),
+                new Claim(JwtRegisteredClaimNames.Jti,
+                    Guid.NewGuid()
+                        .ToString())
```
If you're chopping method parameters please use the following style
```
new Claim(
    JwtRegisteredClaimNames.Sub,
    userInfo.UserName
)
```
But in most cases if it fits in one line (120 characters) then you don't have to chop parameters at all

andrei-kruhliakou-itechart commented on iTechArt.CinemaWebApp.API/Application/Wrappers/Response.cs:
```diff
@@ -0,0 +1,26 @@
+        public string Message { get; set; }
+        public T Data { get; set; }
+
+        public Response()
+        {
+        }
+
```
What would you use parameterless constructor for?

andrei-kruhliakou-itechart commented on iTechArt.CinemaWebApp.API/Controllers/FilmsController.cs:
```diff
@@ -62,12 +63,12 @@ public async Task<ActionResult<Film>> CreateFilm(Film film)
                 Description = film.Description,
-                PosterUrl = (!string.IsNullOrEmpty(film.PosterUrl)) ? film.PosterUrl : "",
-                BannerUrl = (!string.IsNullOrEmpty(film.BannerUrl)) ? film.BannerUrl : ""
+                PosterUrl = !string.IsNullOrEmpty(film.PosterUrl) ? film.PosterUrl : "",
```
It is preferable to use string.Empty over "".

andrei-kruhliakou-itechart commented on iTechArt.CinemaWebApp.API/Models/Policies.cs:
```diff
@@ -0,0 +1,20 @@
+
+        public static AuthorizationPolicy AdminPolicy()
+        {
+            return new AuthorizationPolicyBuilder().RequireAuthenticatedUser().RequireRole(Admin).Build();
```
Please place chaining method calls from the new line
```suggestion
            return new AuthorizationPolicyBuilder()
                .RequireAuthenticatedUser()
                .RequireRole(Admin)
                .Build();
```

bagatka commented:
I've checked "files changed".
Seems like all done.
